improving tree rendering

akosyakov opened the pull request:
This PR improves the tree rendering:
- batching decoration updates, before on each keystroke a tree was re-rendered several times for each decorator
- rendering a tree only if it is attached and visible, before all trees were rerendered on each keystroke regardless of whether they are visible
- rendering only visible part of a tree (virtualization), before all nodes were rendered always

It improves overall performance since the main thread can be busy with the processing of keyboard events and requests from language servers instead of delaying it by rerendering huge trees, e.g. issues like https://github.com/theia-ide/theia/issues/2220 were caused by it.

In order to reproduce and verify, I do the following:
- open a lot of folders in node_modules and packages
- open several files, like 4 or 5
- type `\tHello World`

Measurement before:
<img width="1628" alt="before" src="https://user-images.githubusercontent.com/3082655/42409196-f8f06ce4-81d6-11e8-876b-b5b0c70834cd.png">

You can see big boxes with red corners. It is tree rendering, all other handlers (keyboard, requests) are squeezed between them.

Measurement after:
<img width="1628" alt="after" src="https://user-images.githubusercontent.com/3082655/42409204-2fc45e10-81d7-11e8-8032-c60b48be9650.png">

Big boxes are gone. Editor and trees shold feel snappier.

marcdumais-work commented:
Potential bug: the file/ws open dialog seems to be missing its file tree:
![image](https://user-images.githubusercontent.com/25749063/42409926-5f85b36a-81af-11e8-84ee-57c14428fdc7.png)

marcdumais-work commented:
It will be interesting to measure if and how much this PR improves the performance when very many error markers make the Problems view very slow: https://github.com/theia-ide/theia/issues/2060

akosyakov commented:
@marcdumais-work I've pushed again, the issue was caused partly by me and partly by missing `super.init`: https://github.com/theia-ide/theia/pull/2305/files#diff-4f77795dc1c1bfecb7f6e3e376cda8ddR30. Not it should be good and #1751 + #1692 fixed as well.

Problem view with a lot of markers still can freeze UI during creating a list of visible nodes from a tree :( After creation, it is very smooth :). It should be possible to create a list lazily as a user scrolls the tree to improve it further.

marcdumais-work commented:
Thanks @akosyakov . I just tried and with the latest commit, this PR doesn't build for me anymore.
```
@theia/core: [lint] theiaext lint exited with code 0
@theia/core: [compile] src/browser/tree/tree-iterator.spec.ts(60,28): error TS2554: Expected 2 arguments, but got 1.
@theia/core: [compile] src/browser/tree/tree-iterator.spec.ts(73,28): error TS2554: Expected 2 arguments, but got 1.
@theia/core: [compile] src/browser/tree/tree-iterator.spec.ts(93,73): error TS2345: Argument of type '{ pruneCollapsed: true; }' is not assignable to parameter of type 'Options | undefined'.
@theia/core: [compile]   Type '{ pruneCollapsed: true; }' is not assignable to type 'Options'.
@theia/core: [compile]     Property 'pruneSiblings' is missing in type '{ pruneCollapsed: true; }'.
@theia/core: [compile] src/browser/tree/tree-iterator.spec.ts(99,28): error TS2554: Expected 2 arguments, but got 1.
@theia/core: [compile] src/browser/tree/tree-model.ts(282,55): error TS2345: Argument of type '{ pruneCollapsed: true; }' is not assignable to parameter of type 'Options | undefined'.
@theia/core: [compile]   Type '{ pruneCollapsed: true; }' is not assignable to type 'Options'.
@theia/core: [compile]     Property 'pruneSiblings' is missing in type '{ pruneCollapsed: true; }'.
@theia/core: [compile] theiaext compile exited with code 2
@theia/core: error An unexpected error occurred: "Command failed.
```

akosyakov commented:
@marcdumais-work pull again, sorry did not check test compilation first :)

marcdumais-work commented:
no prob :)

akosyakov commented:
@svenefftinge try again, both issues should be fixed

marcdumais-work commented:
I also see that ```.yml``` files have some coloring but not a lot - probably buggy. 

A big one that's missing is JSON files. But I think there's a plan to have a LS for that, where we would probably store the TextMate grammar.

akosyakov commented:
@marcdumais-work I've added one more commit to precompute next and previous siblings of a node. It was actually a bottleneck in creating a tree for the problem view. I don't see the problem view as a hot spot in the profiler anymore. It moved to the outline, in cases with a single huge file the outline is super greedy and hangs UI. But if you have a lot of problems with small files UI should not freeze anymore, as in https://github.com/theia-ide/theia/issues/2060

akosyakov commented:
@marcdumais-work @svenefftinge I've reimplemented the outline. Before it was doing incredible amount of [repetitive traversing of all symbols](https://github.com/theia-ide/theia/blob/f517558f0757e8aa9355bdd55edb349916b13a75/packages/monaco/src/browser/monaco-outline-contribution.ts#L198-L231). Look for all `filter`, `map`, `forEach` through all symbols and also realize it is done for each symbol recursively.

I've replaced it with 3 passes:
- sorting
- creating nodes and indexing them by a symbol name (we can do it since it is a prerequisite for the containment relationship):
https://github.com/theia-ide/theia/blob/3c381c2533098154d60dc1ded9c1e2411261aff2/packages/monaco/src/browser/monaco-outline-contribution.ts#L131-L138
- establishing the containment relationship between nodes based on the index:
https://github.com/theia-ide/theia/blob/3c381c2533098154d60dc1ded9c1e2411261aff2/packages/monaco/src/browser/monaco-outline-contribution.ts#L140-L157

After this change, the outline is not even the hot spot for 20k+ files. I've tested with [checker.ts](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/checker.ts). I am capable of editing it without UI freezing with the outline, the problem and the navigator views opened. Although, the problem and the outline views needs significant time to catch up. The most of the time is spent on TextMate tokenizing, but it is already unrelated to trees.

Also, since computation of the outline tree is detached from the outline widget, it was computed even if the widget is not present. I fixed it, so if one has issues with the outline he can get rid of them by closing the widget.

AlexTugarev commented on packages/monaco/src/browser/monaco-outline-contribution.ts:
```diff
@@ -32,27 +32,28 @@ import debounce = require('lodash.debounce');
+            if (open) {
+                this.toDisposeOnClose.push(this.toDisposeOnEditor);
+                this.toDisposeOnClose.push(DocumentSymbolProviderRegistry.onDidChange(
+                    debounce(() => this.updateOutline())
```
@akosyakov, I'm confused that there is no delay here. AFAIK `DocumentSymbolProviderRegistry` might flood us here with events. I think that debounce should have a delay of some ms and also it should be called with `{ leading: true }`, so that we start seeing something right after opening the editor.

akosyakov replied:
I have not changed it, it was like that before. Could you check and file an issue if it is true?

marcdumais-work commented:
> I've added one more commit to precompute next and previous siblings of a node. It was actually a bottleneck in creating a tree for the problem view.  I don't see the problem view as a hot spot in the profiler anymore.

Awesome, thanks @akosyakov !
